src/helper.cc

```diff
@@ -891,38 +890,30 @@ Helper::Client::handleFewerServers(const bool madeProgress)
 }
 
 void
-Helper::Session::HelperServerClosed(Session * const srv)
+Helper::Client::serverClosed(SessionBase * const srv)
 {
-    const auto hlp = srv->parent;
-
     bool needsNewServers = false;
-    hlp->handleKilledServer(srv, needsNewServers);
+    handleKilledServer(srv, needsNewServers);
     if (needsNewServers) {
         debugs(80, DBG_IMPORTANT, "Starting new helpers");
-        helperOpenServers(hlp);
+        openServers();
     }
+}
 
+void
+Helper::Session::HelperServerClosed(Session * const srv)
+{
+    srv->parent->serverClosed(srv);
     srv->dropQueued();
-
     delete srv;
 }
 
-// XXX: Almost duplicates Helper::Session::HelperServerClosed() because helperOpenServers() is not a virtual method of the `Helper::Client` class
-// TODO: Fix the `Helper::Client` class hierarchy to use virtual functions.
+// XXX: duplicates Helper::Session::HelperServerClosed()
```

Comment:
Since official dropQueued() is already virtual, I do not see why helper_stateful_server::HelperServerClosed() should continue to duplicate Helper::Session::HelperServerClosed() in this PR. Unless I am missing something, please refactor to move/merge both methods into a single Helper::SessionBase::HandleClosure(SessionBase*).

I do not think this PR should de-duplicate the code registering these closure handlers.

Reply:
We cannot do this because SessionBase does not have `parent` member - it is specified in its kids (with the corresponding Client type).

Reply:
We should add a pure virtual SessionBase::helper() returning a non-constant reference to Helper::Client. However, after looking through the existing code, I can tell that adding that method should trigger many out-of-scope changes. Let's do that separately, as a follow up PR.

```suggestion
// XXX: Essentially duplicates Helper::Session::HelperServerClosed() until we add SessionBase::helper().
```

Reply:
Done.